scanpy/preprocessing/_scrublet/__init__.py

```diff
@@ -429,38 +410,31 @@ def _scrublet_call_doublets(
     # Ensure normalised matrix sparseness as Scrublet does
     # https://github.com/swolock/scrublet/blob/67f8ecbad14e8e1aa9c89b43dac6638cebe38640/src/scrublet/scrublet.py#L100
 
-    scrub._E_obs_norm = sparse.csc_matrix(adata_obs.X)
-    scrub._E_sim_norm = sparse.csc_matrix(adata_sim.X)
+    scrub._counts_obs_norm = sparse.csc_matrix(adata_obs.X)
+    scrub._counts_sim_norm = sparse.csc_matrix(adata_sim.X)
 
     scrub.doublet_parents_ = adata_sim.obsm["doublet_parents"]
 
     # Call scrublet-specific preprocessing where specified
 
     if mean_center and normalize_variance:
-        sl.pipeline_zscore(scrub)
+        pipeline.zscore(scrub)
     elif mean_center:
-        sl.pipeline_mean_center(scrub)
+        pipeline.mean_center(scrub)
     elif normalize_variance:
-        sl.pipeline_normalize_variance(scrub)
+        pipeline.normalize_variance(scrub)
 
     # Do PCA. Scrublet fits to the observed matrix and decomposes both observed
     # and simulated based on that fit, so we'll just let it do its thing rather
     # than trying to use Scanpy's PCA wrapper of the same functions.
 
     if mean_center:
         logg.info("Embedding transcriptomes using PCA...")
-        # Sklearn PCA doesn't like matrices, so convert to arrays
-        if isinstance(scrub._E_obs_norm, np.matrix):
-            scrub._E_obs_norm = np.asarray(scrub._E_obs_norm)
-        if isinstance(scrub._E_sim_norm, np.matrix):
-            scrub._E_sim_norm = np.asarray(scrub._E_sim_norm)
-        sl.pipeline_pca(
-            scrub, n_prin_comps=n_prin_comps, random_state=scrub.random_state
-        )
+        pipeline.pca(scrub, n_prin_comps=n_prin_comps, random_state=scrub._random_state)
     else:
         logg.info("Embedding transcriptomes using Truncated SVD...")
-        sl.pipeline_truncated_svd(
-            scrub, n_prin_comps=n_prin_comps, random_state=scrub.random_state
+        pipeline.truncated_svd(
+            scrub, n_prin_comps=n_prin_comps, random_state=scrub._random_state
```

Comment:
See other comment, but I don't really get why `self.random_state` can't be assigned in `__post_init__`

Reply:
I haven’t found a better dataclass pattern for being able to pass types `X | Y` to the initializer while the corresponding field is of type `X` always.

There‘s also no way I’m not using a dataclass with the current design, because that’s the only way to handle that number of fields without having everything duplicated between `__init__` and the field types.

Reply:
Ok, I guess there's no way around this then.  We want to make `random_state` public via `__init__` but then need to use it again in a different form.   Unfortunate.  I guess, though, `_random_state` is more of a generator though?  Could call it `rng`?  But fine leaving this